From c15a25a61d64efe8f7b29c1e2d0a96885835ec59 Mon Sep 17 00:00:00 2001
From: Mike Frysinger <vapier@gentoo.org>
Date: Wed, 21 Nov 2018 11:00:54 +0100
Subject: [PATCH] use awk/sed instead of perl for creating header files

More systems should have awk/sed than perl.

Signed-off-by: Mike Frysinger <vapier@gentoo.org>

Forward ported from libcap-2.22 to libcap-2.26
and incorporated the gperf-3.1 fix provided by Mike Gilbert
<floppym@gentoo.org>

Signed-off-by: Lars Wendler <polynomial-c@gentoo.org>
---
 libcap/Makefile | 21 +++++++++++++++++----
 1 file changed, 17 insertions(+), 4 deletions(-)

diff --git a/libcap/Makefile b/libcap/Makefile
index 77b26c5..a3f0507 100644
--- a/libcap/Makefile
+++ b/libcap/Makefile
@@ -18,6 +18,8 @@ MAJLIBNAME=$(LIBNAME).$(VERSION)
 MINLIBNAME=$(MAJLIBNAME).$(MINOR)
 GPERF_OUTPUT = _caps_output.gperf
 CFLAGS += -fPIC
+AWK = awk
+SED = sed
 
 all: $(MINLIBNAME) $(STALIBNAME) libcap.pc
 
@@ -42,11 +44,22 @@ cap_names.h: _makenames
 	./_makenames > cap_names.h
 
 $(GPERF_OUTPUT): cap_names.list.h
-	perl -e 'print "struct __cap_token_s { const char *name; int index; };\n%{\nconst struct __cap_token_s *__cap_lookup_name(const char *, size_t);\n%}\n%%\n"; while ($$l = <>) { $$l =~ s/[\{\"]//g; $$l =~ s/\}.*// ; print $$l; }' < $< | gperf --ignore-case --language=ANSI-C --readonly --null-strings --global-table --hash-function-name=__cap_hash_name --lookup-function-name="__cap_lookup_name" -c -t -m20 $(INDENT) > $@
+	(printf "%b" "struct __cap_token_s { const char *name; int index; };\n%%\n"; \
+	$(SED) -e 's:["{}]::g' -e 's:,$$::' $<) | \
+	gperf \
+		--ignore-case \
+		--language=ANSI-C \
+		--includes \
+		--readonly \
+		--null-strings \
+		--global-table \
+		--hash-function-name=__cap_hash_name \
+		--lookup-function-name="__cap_lookup_name" \
+		-c -t -m20 $(INDENT) > $@
 
-cap_names.list.h: Makefile $(KERNEL_HEADERS)/linux/capability.h
-	@echo "=> making $@ from $(KERNEL_HEADERS)/linux/capability.h"
-	perl -e 'while ($$l=<>) { if ($$l =~ /^\#define[ \t](CAP[_A-Z]+)[ \t]+([0-9]+)\s+$$/) { $$tok=$$1; $$val=$$2; $$tok =~ tr/A-Z/a-z/; print "{\"$$tok\",$$val},\n"; } }' $(KERNEL_HEADERS)/linux/capability.h | fgrep -v 0x > $@
+cap_names.list.h: $(KERNEL_HEADERS)/linux/capability.h Makefile
+	@echo "=> making $@ from $<"
+	$(AWK) '($$0 ~ /^#define[[:space:]]+CAP[_A-Z]+[[:space:]]+[0-9]+[[:space:]]*$$/) { printf "{\"%s\",%s},\n", tolower($$2), $$3 }' $< > $@
 
 $(STALIBNAME): $(OBJS)
 	$(AR) rcs $@ $^
-- 
2.20.0.rc0

